### PR TITLE
Fix fcode

### LINKED ISSUE #12267: Codegen: sign function in Fortran
The Fortran code generated by Sympy for the sign function is not a valid Fortran syntax.

With Sympy 1.0 and Python 3.6:
```python
In [1]: import sympy as sp

In [2]: from sympy.abc import x

In [3]: sp.fcode(sp.sign(x))
Out[3]: '      sign(x)'
```
(The same behavior is obtained with `sympy.utilities.codegen`.)

The syntax of the sign function in Fortran is given [here](https://gcc.gnu.org/onlinedocs/gfortran/SIGN.html).

I guess it would be preferable to generate something like `sign(1, x)` or `sign(1d0, x)` depending of the context. I'm not sure how to get the same behavior than Sympy for `x=0`, but at least it would be compilable.

### PR DESCRIPTION
**Pull request by SatyaPrakashDwibedi**
Fixes #12267 @bjodah

### CODE REVIEW

File: sympy/printing/fcode.py [hunk 1]
```diff
@@ -106,6 +106,15 @@ def _get_statement(self, codestring):
 
     def _get_comment(self, text):
         return "! {0}".format(text)
+    
+    #issue 12267
+    def _print_sign(self,func):
+        if func.args[0].is_integer:
+            return "merge(0, isign(1, {0}), {0} == 0)".format(self._print(func.args[0]))
+        elif func.args[0].is_complex:
+            return "merge( cmplx(0d0, 0d0), {0}/abs({0}), abs({0}) == 0d0)".format(self._print(func.args[0]))
```
Comment (bjodah) [inline sympy/printing/fcode.py]: Why the extra space after `merge(`?

**Comment by bjodah** [comment]
@SatyaPrakashDwibedi if this is your current suggestion you can close #12298 
You really should run the relevant tests locally since it only takes a fraction of second (see my comments in your other PR and the issue for how to run tests).

File: sympy/printing/fcode.py (hunk 1 above)
Comment (SatyaPrakashDwibedi) [inline sympy/printing/fcode.py]: @Franjov mentioned so in comment

Comment (bjodah) [inline sympy/printing/fcode.py]: I believe he meant that there was a discrepancy between your test and your print method. You can remove the space.

**Comment by bjodah** [comment]
So I wrote a small test program:

```
program ex4
implicit none
integer, parameter :: N=3
integer, parameter :: dp = kind(1d0)
complex(dp) :: A(N)
integer :: i
A = [(2, 3), (0, 0), (1, 1)]
do i=1,N
   print *, i
   print *, merge(cmplx(0d0, 0d0, kind(1d0)), A(i)/abs(A(i)), abs(A(i)) == 0d0)
end do
end program ex4
```
run by e.g. issuing:
```
$ gfortran -Wall -Werror -std=f2003 ex4.f90 && ./a.out
```
I think this is the preferred form (I believe people use 128 bit complex types since the FPUs in modern processors are optimized for 64 bit IEEE 754 numbers), what do you think @Franjov?

**Comment by SatyaPrakashDwibedi** [comment]
@bjodah I ran the test it did not show any error

File: sympy/printing/fcode.py [hunk 2]
```diff
@@ -106,6 +106,15 @@ def _get_statement(self, codestring):
 
     def _get_comment(self, text):
         return "! {0}".format(text)
+    
+    #issue 12267
+    def _print_sign(self,func):
+        if func.args[0].is_integer:
+            return "merge(0, isign(1, {0}), {0} == 0)".format(self._print(func.args[0]))
+        elif func.args[0].is_complex:
+            return "merge(cmplx(0d0, 0d0), {0}/abs({0}), abs({0}) == 0d0)".format(self._print(func.args[0]))
```
Comment (bjodah) [inline sympy/printing/fcode.py]: Let's add a third argument to cmplx (`kind({0})`), that way the user won't loose precision involuntarely.

File: sympy/printing/fcode.py (hunk 1 above)
Comment (Franjov) [inline sympy/printing/fcode.py]: I confirm

**Comment by Franjov** [comment]
@bjodah Interestingly I notice only now that `cmplx` return a standard precision complex number (that is a couple of standard real numbers) whatever the type of the arguments. It would mean that `cmplx(0d0, 0d0)` could be replaced by `cmplx(0, 0)`.

It seems that the standard behavior of Sympy is to use this level of precision in Fortran code:
```python
In [2]: sp.fcode(1 + sp.I)
Out[2]: '      cmplx(1,1)'

In [3]: sp.fcode(1.0 + 1.0*sp.I)
Out[3]: '      cmplx(1.0d0,1.0d0)'
```
So I guess it is better to stick with it in the sign function.

Although a finer management of Fortran types would make a nice future improvement of `fcode`.

**Comment by bjodah** [comment]
@Franjov I didn't know that either before gfortran warned me about it. I think you are right about deferring 3rd argument fixes to `cmplx` to a later PR. Would you mind opening an issue for it?

File: sympy/printing/fcode.py (hunk 2 above)
Comment (SatyaPrakashDwibedi) [inline sympy/printing/fcode.py]: `'      merge(cmplx(0d0, 0d0, kind(1d0)), x/abs(x), abs(x) == 0d0)'`
or
`'      merge(cmplx(0d0, 0d0, kind(x)), x/abs(x), abs(x) == 0d0)'`
which one is required

Comment (bjodah) [inline sympy/printing/fcode.py]: Sorry, for me changing my mind. But @Franjov pointed out that we are already using the lower precision form om `cmplx` in other places in `fcode`. So you can disregard that comment. If tests passes now this is good to merge without any further changes.

**Comment by Franjov** [comment]
I will think a bit about this question of precision and open an issue in the next days.

File: sympy/printing/fcode.py (hunk 2 above)
Comment (SatyaPrakashDwibedi) [inline sympy/printing/fcode.py]: It is showing trailing white space error again but it is not showing in debugger.
How can I make sure that all the trailing white spaces are removed?

Comment (bjodah) [inline sympy/printing/fcode.py]: I think you fixed it in your last commits, I don't get an error for those but I got it for the earlier ones:
https://asciinema.org/a/0umymdojl3z01mpe7ez0tfwso

**Comment by SatyaPrakashDwibedi** [comment]
@bjodah it passed what should I do now

**Comment by bjodah** [comment]
Good work! This is done for now. Merged.

**Pull request merged**